src/mlpack/core/data/imputer.hpp

```diff
@@ -52,93 +51,47 @@ class Imputer
   * @param dimension.
   * @param transpose.
   */
-  void Impute(const MatType &input,
-              MatType &output,
-              const std::string &targetValue,
+  void Impute(const arma::Mat<T>& input,
+              arma::Mat<T>& output,
+              const std::string& missingValue,
               const size_t dimension)
   {
-    // find mapped value inside current mapper
-    auto mappedValue = mapper.UnmapValue(targetValue, dimension);
-
-    if(transpose)
-    {
-      for (size_t i = 0; i < input.n_rows; ++i)
-      {
-        if (input(dimension, i) == mappedValue)
-        {
-          // users can specify the imputation strategies likes
-          // mean, mode, etc using the class'es template parameter: Strategy.
-          strat.template Impute<MatType>(input, output, dimension, i, transpose);
-        }
-      }
-    }
-    else
-    {
-      for (size_t i = 0; i < input.n_cols; ++i)
-      {
-        if (input(i, dimension) == mappedValue)
-        {
-          strat.template Impute<MatType>(input, output, i, dimension, transpose);
-        }
-      }
-    }
+    T mappedValue = static_cast<T>(mapper.UnmapValue(missingValue, dimension));
+    strategy.Apply(input, output, mappedValue, dimension, transpose);
   }
 
   /**
   * This overload of Impute() lets users to define custom value that
   * can be replaced with the target value.
   */
-  template <typename T>
-  void Impute(const arma::Mat<T> &input,
-              arma::Mat<T> &output,
-              const std::string &targetValue,
-              const T &customValue,
+  void Impute(const arma::Mat<T>& input,
```

Comment:
Maybe we could std::enable_if to specify which type of imputation are support by this function?Or static_assert?If it works, I think static_assert is a better choice, since the error messages should be more human friendly.